src/utils/callDockerode.ts

```diff
@@ -24,32 +27,48 @@ export async function callDockerode<T>(dockerodeCallback: () => T): Promise<T> {
         }
     });
 
-    return callDockerodeAsync(async () => p);
+    return callDockerodeAsync(async () => p, context);
 }
 
-export async function callDockerodeAsync<T>(dockerodeAsyncCallback: () => Promise<T>): Promise<T> {
-    // If running tests, don't refresh Dockerode (some tests override Dockerode)
-    if (!ext.runningTests) {
-        const { Changed: contextChanged } = await dockerContextManager.getCurrentContext();
-        if (contextChanged) {
-            /* eslint-disable no-unused-expressions */
-            // This will cause any still-awaiting promises to reject with UserCancelledError, and then cps will be recreated
-            cps?.cancel();
-            cps?.dispose();
-            cps = undefined;
-            /* eslint-enable no-unused-expressions */
-
-            await refreshDockerode();
+export async function callDockerodeAsync<T>(dockerodeAsyncCallback: () => Promise<T>, context?: IActionContext): Promise<T> {
+    const timeoutPromiseSource = new CancellationPromiseSource(Error, localize('vscode-docker.utils.dockerode.timeout', 'Request timed out.'));
```

Comment:
I think we need another utility class for this, e.g. `TimeoutPromiseSource`, in `promiseUtils`. The timer handling logic could go there. This utility would take a timeout and a lambda that would be called if the timeout elapses. Disposing would stop the internal timer automatically.

Having such utility class would simplify the code in `callDockerode`